qubes/mgmt.py

```diff
@@ -144,37 +133,96 @@ def vm_list(self, untrusted_payload):
+        # explicit list to be sure that it matches protocol spec
+        if isinstance(property_def, qubes.vm.VMProperty):
+            property_type = 'vm'
+        elif property_def.type is int:
+            property_type = 'int'
+        elif property_def.type is bool:
+            property_type = 'bool'
+        elif self.arg == 'label':
+            property_type = 'label'
+        else:
+            property_type = 'str'
+
         try:
             value = getattr(self.dest, self.arg)
         except AttributeError:
-            return 'default=True '
+            return 'default=True type={} '.format(property_type)
         else:
-            return 'default={} {}'.format(
+            return 'default={} type={} {}'.format(
                 str(self.dest.property_is_default(self.arg)),
-                self.repr(value))
+                property_type,
+                str(value) if value is not None else '')
 
     @asyncio.coroutine
-    def vm_property_help(self, untrusted_payload):
+    def vm_property_set(self, untrusted_payload):
+        assert self.arg in self.dest.property_list()
+
+        property_def = self.dest.property_get_def(self.arg)
+        # do not treat type='str' as sufficient validation
+        if isinstance(property_def, qubes.vm.VMProperty):
+            try:
+                untrusted_vmname = untrusted_payload.decode('ascii')
+            except UnicodeDecodeError:
+                raise qubes.exc.QubesValueError
+            qubes.vm.qubesvm.validate_name(self.dest, property_def,
+                untrusted_vmname)
+            try:
+                newvalue = self.app.domains[untrusted_vmname]
+            except KeyError:
+                raise qubes.exc.QubesPropertyValueError(
+                    self.dest, property_def, untrusted_vmname)
+        elif property_def.type is not None and property_def.type is not str:
+            if property_def.type is bool:
+                try:
+                    untrusted_value = untrusted_payload.decode('ascii')
+                except UnicodeDecodeError:
+                    raise qubes.exc.QubesValueError
+                newvalue = qubes.property.bool(None, None, untrusted_value)
+            else:
+                try:
+                    newvalue = property_def.type(untrusted_payload)
+                except ValueError:
+                    raise qubes.exc.QubesValueError
+        else:
+            # other types (including explicitly defined 'str') coerce to str
+            # with limited characters allowed
+            try:
+                untrusted_value = untrusted_payload.decode('ascii')
+            except UnicodeDecodeError:
+                raise qubes.exc.QubesValueError
+            allowed_set = string.printable
+            if not all(x in allowed_set for x in untrusted_value):
+                raise qubes.exc.QubesValueError(
+                    'Invalid characters in property value')
+            newvalue = untrusted_value
```

Comment:
@jpouellet Yes, this is already done. May I point you in those directions:
- https://qubes-core-admin.readthedocs.io/en/latest/qubes.html#properties
- https://github.com/QubesOS/qubes-core-admin/blob/3e0f225/qubes/__init__.py#L130
- https://github.com/QubesOS/qubes-core-admin/blob/3e0f225/qubes/vm/qubesvm.py#L355-L489 with specific focus on `setter=`

The question is, should we provide some additional (coarse) validation to all properties that cannot be overridden with specific intent to eliminate some basic attacks, possibly encoding-oriented. The problem is, when I wrote `setter=`, I thought it would be used from dom0 and we wouldn't supply it with untrusted data (possibly also malformed). If we'd like to repurpose that, we should look at it once again, rename function signatures to `untrusted_*` (possibly enforcing the name as keyword-only argument, since we have python3).